[config]Static routes to config_db

d-dashkov opened the pull request:
**What I did**
fixes https://github.com/Azure/sonic-buildimage/issues/6997
Added process for writing static routes to config_db.

fixes https://github.com/Azure/sonic-buildimage/issues/7261
Added interface validation before creating route. 

**How I did it**
Refactored add/del route functions. Wrote generic function that returns a vtysh command and an entry for config_db with name.
```
sudo config route add prefix 1.2.3.4/32 nexthop 30.0.0.5
sudo config route add prefix 1.3.3.4/32 nexthop 30.0.0.5
sudo config route add prefix 1.4.3.4/32 nexthop 30.0.0.5
sudo config route add prefix vrf Vrf-RED 1.5.3.4/32 nexthop 30.0.0.5
sudo config save -y
sudo reboot
```
Instead of running the vtysh command, it creates config_db entries that will be loaded by the ffr daemon:
```
    "STATIC_ROUTE": {
        "1.2.3.4/32": {
            "nexthop": "30.0.0.5"
        },
        "1.3.3.4/32": {
            "nexthop": "30.0.0.5"
        },
        "1.4.3.4/32": {
            "nexthop": "30.0.0.5"
        },
        "Vrf-RED|1.5.3.4/32": {
            "nexthop": "30.0.0.5"
        }
    }
```

prsunny commented:
Here, the user is expected to run `route update` after every reboot?

d-dashkov commented:
@prsunny 
I don't think so, user shouldn't run this command manually after reboot. But I really don't know where this update function can be called, because it should be executed after the RIFs are generated, and running it in the startup script shouldn't work.

prsunny commented:
> @prsunny
> I don't think so, user shouldn't run this command manually after reboot. But I really don't know where this update function can be called, because it should be executed after the RIFs are generated, and running it in the startup script shouldn't work.

ok, so how is it expected to be restored from config_db after reboot?

prsunny commented:
Can you add unit-tests as well?

prsunny commented:
Why is the schema showing nexthop in key?
```
       "1.2.3.4/32|30.0.0.5": {
            "nexthop": "30.0.0.5",
            "prefix": "1.2.3.4/32"
        },
```

caizhenghui-juniper commented:
And also why do we need prefix in the value field? It's already part of the key... 
For static route table the scheme is prefix as key and nexthop as field as the simplest case..

caizhenghui-juniper commented:
we already have frrcfgd monitoring config db, so the part of programming frr cfg is not necessary as long as config_db is updated, isn't it?

d-dashkov commented:
@prsunny @caizhenghui-juniper
I updated the record name with a prefix as the key and nexthop as the value, and also added base UT. But I cannot find frrcfgd in the repositories, can you tell me where it is?

caizhenghui-juniper commented:
Hi, Dmytro,

Here it is: https://github.com/Azure/sonic-buildimage/tree/master/src/sonic-frr-mgmt-framework/frrcfgd
One more issue: in swss, we do require VRF name conforms with prefix Vrf, so please correct that in VRF test case accordingly. Small problem 😊
And the schema defined in frrcfgd for vrf is using vrf_name instead of vrf.
Plus, are you still setting prefix as field? At line 3248?

Thanks

Zhenghui

From: Dmytro ***@***.***>
Reply-To: Azure/sonic-utilities ***@***.***>
Date: Wednesday, March 31, 2021 at 1:52 PM
To: Azure/sonic-utilities ***@***.***>
Cc: Zhenghui Cai ***@***.***>, Mention ***@***.***>
Subject: Re: [Azure/sonic-utilities] [config]Static routes to config_db (#1534)

[External Email. Be cautious of content]


@prsunny<https://github.com/prsunny> @caizhenghui-juniper<https://github.com/caizhenghui-juniper>
I updated the record name with a prefix as the key and nexthop as the value, and also added base UT. But I cannot find frrcfgd in the repositories, can you tell me where it is?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/Azure/sonic-utilities/pull/1534#issuecomment-811288563>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AKDUQJVEG3JZD3HB5MP773DTGNOOTANCNFSM42CKDI4Q>.


Juniper Business Use Only

caizhenghui-juniper commented:
and can you put support for ECMP too? With more than 1 next hops for the static routes?

shi-su commented on tests/static_routes_test.py:
```diff
@@ -0,0 +1,100 @@
+        ["prefix", "3.2.3.4/32", "nexthop", "vrf", "Vrfred", "30.0.0.6"], obj=obj)
+        print(result.exit_code, result.output)
+        assert ('3.2.3.4/32') in db.cfgdb.get_table('STATIC_ROUTE')
+        assert db.cfgdb.get_entry('STATIC_ROUTE', '3.2.3.4/32') == {"nexthop": "30.0.0.6", "nexthop_vrf": "Vrfred"}
```
I think we need to follow a certain schema design for the STATIC_ROUTE table. In https://github.com/Azure/SONiC/pull/585, there is a schema for the table but it seems different here.

d-dashkov commented:
I'm not sure if this is a bug or not, but after reboot, the frr daemon does not start with the system, and also after manual start, the daemon sees a static route, but does not add it to frr cfg

**Before reboot**
```
sudo config route add prefix 1.2.3.4/32 nexthop 30.0.0.5
```
```
"STATIC_ROUTE": {
	"1.2.3.4/32": {
		"nexthop": "30.0.0.5"
	}
}


admin@sonic:~$ redis-cli -n 4 keys "*" | grep STATIC_ROUTE
STATIC_ROUTE|1.2.3.4/32
admin@sonic:~$ redis-cli -n 4 hgetall "STATIC_ROUTE|1.2.3.4/32"
1) "nexthop"
2) "30.0.0.5"
```
```
admin@sonic:~$ show ip route
Codes: K - kernel route, C - connected, S - static, R - RIP,
       O - OSPF, I - IS-IS, B - BGP, E - EIGRP, N - NHRP,
       T - Table, v - VNC, V - VNC-Direct, A - Babel, D - SHARP,
       F - PBR, f - OpenFabric,
       > - selected route, * - FIB route, q - queued, r - rejected, b - backup

S>* 0.0.0.0/0 [200/0] via 192.168.111.3, eth0, weight 1, 00:05:24
S>  1.2.3.4/32 [1/0] via 30.0.0.7 (recursive), weight 1, 00:00:06
C>* 10.1.0.1/32 is directly connected, Loopback0, 00:05:07
C>* 192.168.111.0/24 is directly connected, eth0, 00:05:25
```

**After reboot**

Syslog
```
May  7 21:07:06.926529 sonic DEBUG bgp#/frrcfgd:   STATIC_ROUTE&&1.2.3.4/32 : {'nexthop': '30.0.0.5'}
```
```
admin@sonic:~$ show ip route
Codes: K - kernel route, C - connected, S - static, R - RIP,
       O - OSPF, I - IS-IS, B - BGP, E - EIGRP, N - NHRP,
       T - Table, v - VNC, V - VNC-Direct, A - Babel, D - SHARP,
       F - PBR, f - OpenFabric,
       > - selected route, * - FIB route, q - queued, r - rejected, b - backup

S>* 0.0.0.0/0 [200/0] via 192.168.111.3, eth0, weight 1, 00:05:24
C>* 10.1.0.1/32 is directly connected, Loopback0, 00:05:07
C>* 192.168.111.0/24 is directly connected, eth0, 00:05:25
```
And if I try to change the nexthop on this entry, staticd crashes because it has no previous entry
```
sudo config route add prefix 1.2.3.4/32 nexthop 30.0.0.5
```
```
May  7 21:09:27.483443 sonic DEBUG bgp#/frrcfgd: ----------------------------------
May  7 21:09:27.483607 sonic DEBUG bgp#/frrcfgd:  BGP table handling
May  7 21:09:27.483690 sonic DEBUG bgp#/frrcfgd: ----------------------------------
May  7 21:09:27.483764 sonic DEBUG bgp#/frrcfgd: table : STATIC_ROUTE
May  7 21:09:27.483881 sonic DEBUG bgp#/frrcfgd: key   : 1.2.3.4/32
May  7 21:09:27.483958 sonic DEBUG bgp#/frrcfgd: op    : SET
May  7 21:09:27.484074 sonic DEBUG bgp#/frrcfgd: data  :
May  7 21:09:27.484152 sonic DEBUG bgp#/frrcfgd:         nexthop    - 30.0.0.7
May  7 21:09:27.484225 sonic DEBUG bgp#/frrcfgd:
May  7 21:09:27.484326 sonic INFO bgp#/frrcfgd: value for table STATIC_ROUTE prefix default key 1.2.3.4/32 changed to {'nexthop': (30.0.0.7, UPDATE)}
May  7 21:09:27.484439 sonic INFO bgp#/frrcfgd: Set static IP route for vrf default prefix 1.2.3.4/32
May  7 21:09:27.484525 sonic DEBUG bgp#/frrcfgd: execute command vtysh -c 'configure terminal' -c 'vrf default' -c 'no ip route 1.2.3.4/32  30.0.0.5     ' for table STATIC_ROUTE.
May  7 21:09:27.484646 sonic DEBUG bgp#/frrcfgd: VTYSH CMD: configure terminal daemons: ['staticd']
May  7 21:09:27.484725 sonic DEBUG bgp#/frrcfgd: VTYSH CMD: vrf default daemons: ['staticd']
May  7 21:09:27.485096 sonic DEBUG bgp#/frrcfgd: VTYSH CMD: no ip route 1.2.3.4/32  30.0.0.5 daemons: ['staticd']
May  7 21:09:27.485581 sonic DEBUG bgp#/frrcfgd: [staticd] command return code: 1
May  7 21:09:27.485777 sonic DEBUG bgp#/frrcfgd:
May  7 21:09:27.485883 sonic DEBUG bgp#/frrcfgd: VTYSH CMD: end daemons: ['staticd']
May  7 21:09:27.485995 sonic ERR bgp#/frrcfgd: command execution failure. Command: "vtysh -c 'configure terminal' -c 'vrf default' -c 'no ip route 1.2.3.4/32  30.0.0.5     '"
May  7 21:09:27.486073 sonic ERR bgp#/frrcfgd: failed running FRR command: no ip route 1.2.3.4/32  30.0.0.5
May  7 21:09:27.486161 sonic ERR bgp#/frrcfgd: failed running static route config command
May  7 21:09:27.486235 sonic DEBUG bgp#/frrcfgd: ignore cache update for nexthop because of STAT_FAIL
May  7 21:09:27.486308 sonic INFO bgp#/frrcfgd: delete table row STATIC_ROUTE&&default|1.2.3.4/32 from cache
```

prsunny commented on tests/static_routes_test.py:
```diff
@@ -0,0 +1,119 @@
+        runner = CliRunner()
+        obj = {'config_db':db.cfgdb}
+
+        # config route add prefix 2.2.3.4/32 nexthop 30.0.0.6
```
please update the comment to include vrf

prsunny commented on tests/static_routes_test.py:
```diff
@@ -0,0 +1,119 @@
+        runner = CliRunner()
+        obj = {'config_db':db.cfgdb}
+
+        # config route add prefix 3.2.3.4/32 nexthop 30.0.0.6
```
same here, please update the correct command with vrf

prsunny commented on config/main.py:
```diff
@@ -3153,111 +3196,28 @@ def del_vrf_vni_map(ctx, vrfname):
-    clicommon.run_command(cmd)
+    key, route = cli_sroute_to_config(ctx, command_str)
+    config_db = ctx.obj['config_db']
+    config_db.set_entry("STATIC_ROUTE", key, route)
```
If user update the route, this would overwrite the entry, correct?

prsunny reviewed (commented):
I see that this PR is missing ecmp handling. In the existing code, by calling vtysh, it does handle ECMP. My suggestion is, when user configure the same route prefix with a different nexthop, it should modify the STATIC_ROUTE DB entry instead of overwriting thus supporting ECMP (similar to vtysh).  @shi-su, this should be handled by bgpcfgd.

prsunny commented:
Please update CLI command reference [document ](https://github.com/Azure/sonic-utilities/blob/master/doc/Command-Reference.md)

shi-su commented:
> I see that this PR is missing ecmp handling. In the existing code, by calling vtysh, it does handle ECMP. My suggestion is, when user configure the same route prefix with a different nexthop, it should modify the STATIC_ROUTE DB entry instead of overwriting thus supporting ECMP (similar to vtysh). @shi-su, this should be handled by bgpcfgd.

Yes, the STATIC_ROUTE DB entry needs to be updated when nexthop changes. The bgpcfgd should be able to know the changes and only apply the differences for the same route prefix.

prsunny commented on config/main.py:
```diff
@@ -3153,111 +3197,89 @@ def del_vrf_vni_map(ctx, vrfname):
-            if vrf_name == "":
-                cmd += ' {} nexthop-vrf {}'.format(ip, vrf_name_dst)
+    config_db = ctx.obj['config_db']
+    key, route = cli_sroute_to_config(ctx, command_str)
```
Now that we are writing to DB, lets check the correctness of key (like a valid ip prefix)

d-dashkov replied:
Is it possible to view this PR, it will be much easier to check the ip through this function. This is also my PR, and I can implement any of your suggestions to improve it.

prsunny commented on config/main.py:
```diff
@@ -3153,111 +3197,89 @@ def del_vrf_vni_map(ctx, vrfname):
+        if 'ifname' in current_entry:
+            ifname = current_entry['ifname'].split(',')
+
+        nh_zip = list(itertools.zip_longest(nh, nh_vrf, ifname, fillvalue=''))
```
can you add some comments to understand the logic in this section?

prsunny commented on config/main.py:
```diff
@@ -3153,111 +3197,89 @@ def del_vrf_vni_map(ctx, vrfname):
+            ctx.fail('Not found {} in {}'.format(cmd_tuple, key))
+
+        if len(nh) == 0 or (len(nh) == 1 and nh[0] == ''):
+            config_db.set_entry("STATIC_ROUTE", key, None)
```
if user just deletes the route without mentioning any nexthops, then the route is expected to be fully deleted.

d-dashkov replied:
If the user tries to delete a route without nexthop (for example, 'сonfig del config route del prefix 1.2.3.4/32'), he will receive an error: 'argument is not in pattern'. So if the user has previously added ECMP routes, he must remove them one by one.
This condition checks if the record still has any nexthop, if so, it just updates this record, if it doesn't have the nexthop, it removes the key. Is this the correct behavior or should user be able to delete the entire entry at once?

prsunny replied:
IMO, user should be able to delete the entire route at once.

prsunny commented on tests/static_routes_test.py:
```diff
@@ -0,0 +1,266 @@
+        print(result.exit_code, result.output)
+        assert not ('3.2.3.4/32') in db.cfgdb.get_table('STATIC_ROUTE')
+
+    def test_several_nexthops_static_route(self):
```
please rename it to `test_multiple_nexthops_static_route`

prsunny commented on tests/static_routes_test.py:
```diff
@@ -0,0 +1,266 @@
+        print(result.exit_code, result.output)
+        assert not ('11.2.3.4/32') in db.cfgdb.get_table('STATIC_ROUTE')
+   
+    def test_static_route_ECMP_nextfop_vrf(self):
```
nexthop - typo

prsunny commented on tests/static_routes_test.py:
```diff
@@ -0,0 +1,266 @@
+        assert ('6.2.3.4/32') in db.cfgdb.get_table('STATIC_ROUTE')
+        assert db.cfgdb.get_entry('STATIC_ROUTE', '6.2.3.4/32') == {"nexthop": "30.0.0.6,30.0.0.7"}
+
+        # config route del prefix 6.2.3.4/32 nexthop 30.0.0.7
```
please have a test to del prefix without any nexthops and ensure its removed from DB.

d-dashkov commented:
I found that the daemon cannot create direct routes to interfaces.

When I tried to set it up this way:
```
"10.10.10.0/24": {
	"nexthop": "Ethernet0"
}
```
Daemon returns error: ip in nexthop is invalid

Otherwise:
```
"10.10.10.0/24": {
	"ifname": "Ethernet0"
}
```
Daemon returns error: nexthop ip not found

prsunny commented:
> I found that the daemon cannot create direct routes to interfaces.
> 
> When I tried to set it up this way:
> 
> ```
> "10.10.10.0/24": {
> 	"nexthop": "Ethernet0"
> }
> ```
> 
> Daemon returns error: ip in nexthop is invalid
> 
> Otherwise:
> 
> ```
> "10.10.10.0/24": {
> 	"ifname": "Ethernet0"
> }
> ```
> 
> Daemon returns error: nexthop ip not found

The second case is valid and should be handled by bgpcfgd. @shi-su

shi-su commented:
> > I found that the daemon cannot create direct routes to interfaces.
> > When I tried to set it up this way:
> > ```
> > "10.10.10.0/24": {
> > 	"nexthop": "Ethernet0"
> > }
> > ```
> > 
> > 
> > Daemon returns error: ip in nexthop is invalid
> > Otherwise:
> > ```
> > "10.10.10.0/24": {
> > 	"ifname": "Ethernet0"
> > }
> > ```
> > 
> > 
> > Daemon returns error: nexthop ip not found
> 
> The second case is valid and should be handled by bgpcfgd. @shi-su

bgpcfgd is actually handling the second case:
```
admin@vlab-01:~$ sudo sonic-cfggen -a '{"STATIC_ROUTE":{"10.27.0.0/24": {"ifname": "Ethernet0"}}}' --write-to-db
admin@vlab-01:~$ show ip route 10.27.0.1
Routing entry for 10.27.0.0/24
  Known via "static", distance 1, metric 0
  Last update 00:00:07 ago
    directly connected, Ethernet0 inactive, weight 1
```

prsunny reviewed (approved):
lgtm. The PR lacks the multi-asic namespace support, but that can be planned as a phase 2 activity.

prsunny commented:
@shi-su , could you please review/sign-off?

shi-su commented on config/main.py:
```diff
@@ -3153,111 +3214,110 @@ def del_vrf_vni_map(ctx, vrfname):
+
+        for entry in ['nexthop', 'nexthop-vrf', 'ifname']:
+            if not entry in current_entry:
+                    current_entry[entry] = ''
```
Please remove extra spaces to make the indentation consistent.

d-dashkov replied:
Done

shi-su commented on config/main.py:
```diff
@@ -3153,111 +3214,110 @@ def del_vrf_vni_map(ctx, vrfname):
-            ctx.fail("nexthop is not in pattern!")
-    cmd += '"'
-    clicommon.run_command(cmd)
+                current_entry[entry] += ','
```
Just wanted to confirm, does this command only allow to add one nexthop at a time and something like `config route add prefix 6.2.3.4/32 nexthop "30.0.0.6,30.0.0.7"` is not valid?

prsunny replied:
This is covered in the example and should be supported

shi-su replied:
Then this is a problem. If a command is issued like this `config route add prefix 6.2.3.4/32 nexthop "30.0.0.6,30.0.0.7"`, the nexthop field would add `,30.0.0.6,30.0.0.7` (two additional values), whereas `ifname` and `nexthop-vrf` would only add `,` (one additional value). It follows that the number of values in the fields would mismatch after calling the command.

prsunny replied:
For this command, I wouldn't expect any `ifname `or `nexthop-vrf` attribute since its not mentioned in the CLI. Can you please check `test_multiple_nexthops_static_route(self):.` It looks good to me.

shi-su replied:
`test_multiple_nexthops_static_route` only covers the scenario of adding a new entry, but this part of the code is to add to existing entries. I still doubt if this would work. For example, if we call the command in the following sequence: `config route add prefix 6.2.3.4/32 nexthop "30.0.0.5"` and then ``config route add prefix 6.2.3.4/32 nexthop "30.0.0.6,30.0.0.7"``. The corresponding entry in CONFIG_DB would be `"STATIC_ROUTE":{"6.2.3.4/32": {"nexthop": "30.0.0.5,30.0.0.6,30.0.0.7", "ifname": ",", "nexthop-vrf": ","}}`. This is invalid since `nexthop` has three values and others only have two. I would suggest to fix this and cover the scenario in test.

prsunny replied:
Agree. I thought you were referring to first time route add based on previous comment. @d-dashkov to address this.

d-dashkov replied:
Fixed. Please check test_multiple_nexthops_static_route.
I added alignment for such a case and now ```config route add prefix 6.2.3.4/32 nexthop "30.0.0.6,30.0.0.7"``` creates
 ```{'nexthop': '30.0.0.6,30.0.0.7', 'blackhole': 'false,false', 'distance': '0,0', 'ifname': ',', 'nexthop-vrf': ','}```
Also I added same to ECMP with nexthop vrf - test_multiple_nexthops_with_vrf_static_route

shi-su commented on config/main.py:
```diff
@@ -3153,111 +3214,110 @@ def del_vrf_vni_map(ctx, vrfname):
+
+        config_db.set_entry("STATIC_ROUTE", key, current_entry)
+    else:
+        config_db.set_entry("STATIC_ROUTE", key, route)
```
I think there is an assumption in this design that the static routes are always configured by this CLI command. I think this assumption may not always be true. There are a few scenarios that might be problematic. Here are a few examples:
1. The original `"STATIC_ROUTE":{"1.1.1.1/24": {"nexthop": "2.2.2.2,3.3.3.3"}}` (I know this should never happen with this CLI command only, but it might be valid using some other ways since it is valid according to the schema). If we add an additional nexthop `4.4.4.4`, then with the current logic the entry would become `"STATIC_ROUTE":{"1.1.1.1/24": {"nexthop": "2.2.2.2,3.3.3.3,4.4.4.4", "ifname": ",", "nexthop-vrf": ","}}`. This would be invalid since `nexthop` has three values while others have only two.
2. The original table has some other fields (e.g., distance, blackhole). This command would never modify those fields and the numbers would not match after executing this command.

Similar situations need to be considered in del as well.

prsunny replied:
IMO, scenario 1 could be low prioritized as it wouldn't be a common use-case. @d-dashkov , can you please confirm 2?

d-dashkov replied:
After running this command, the numbers might not match, and to prevent this, I added the creation of these fields and set their default values according to https://github.com/project-arlo/SONiC/blob/1e8551e0ffe6a01ac5a717491814943d37e29ad9/doc/static -route / SONiC_static_route_route .md # mapping-from-table-entry-to-frr-command-arguments, so now user can manually change values in config_db

prsunny commented on config/main.py:
```diff
@@ -3153,111 +3214,110 @@ def del_vrf_vni_map(ctx, vrfname):
+    key, route = cli_sroute_to_config(ctx, command_str)
+
+    # If defined intf name, check if it exists
+    if 'ifname' in route:
```
As commented by @arlakshm for [PR #1535  ](https://github.com/Azure/sonic-utilities/pull/1535#issuecomment-825330038), please check if `ifname `belongs to `VLAN_INTERFACE`, `INTERFACE `and `PORTCHANNEL_INTERFACE`. If not return failure as "IP interface doesn't exist"

shi-su commented on config/main.py:
```diff
@@ -3227,22 +3227,48 @@ def add_route(ctx, command_str):
             ctx.fail('interface {} doesn`t exist'.format(route['ifname']))
 
+    # Alignment in case the command contains several nexthop ip
+    nh_counter = len(route['nexthop'].split(','))
```
Is this assuming that `nexthop` is given by the command? If I use something like `config route add prefix 1.2.3.4/32 dev Ethernet8`, would it do `nh_count = 0`?

d-dashkov replied:
Updated, in ```config route add prefix 1.2.3.4/32 nexthop dev Ethernet8``` case check once, in case with ip check all nexthop ip.

shi-su reviewed (commented):
I do not have any other comments. Please check with other reviewers.

prsunny commented on config/main.py:
```diff
@@ -3153,111 +3214,155 @@ def del_vrf_vni_map(ctx, vrfname):
+    config_db = ConfigDBConnector()
+    config_db.connect()
+    ctx.obj = {}
+    ctx.obj['config_db'] = config_db
```
@arlakshm, can we add multiasic as another enhancement PR?

arlakshm replied:
@prsunny,  Sure.  Can we create an issue or task to track this?

prsunny replied:
Created issue - https://github.com/Azure/sonic-utilities/issues/1608

prsunny commented on config/main.py:
```diff
@@ -3153,111 +3214,155 @@ def del_vrf_vni_map(ctx, vrfname):
+        else:
+            route['distance'] = '0'
+
+        if 'blackhole' in route:
```
We would need to handle one scenario for blackhole routes. If the user configure with "ifname" as "null", could you please add the attribute 'blackhole' as true.

dprital commented:
@prsunny - Can you please add the tag "Required to 202012" ?

haslersn commented:
Is this already available in the prebuilt sonic-aboot-broadcom.swi ? If not, then how to configure (and persist across reboots) static routes?
